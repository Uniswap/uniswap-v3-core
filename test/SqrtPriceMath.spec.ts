import {BigNumber} from 'ethers'
import {ethers} from 'hardhat'
import {SqrtPriceMathTest} from '../typechain/SqrtPriceMathTest'

import {expect} from './shared/expect'
import {encodePriceSqrt, expandTo18Decimals} from './shared/utilities'

describe.only('SqrtPriceMath', () => {
  let sqrtPriceMath: SqrtPriceMathTest
  before(async () => {
    const priceMathTestFactory = await ethers.getContractFactory('SqrtPriceMathTest')
    sqrtPriceMath = (await priceMathTestFactory.deploy()) as SqrtPriceMathTest
  })

  describe('#getPriceAfterSwap', () => {
    it('fails if price is zero', async () => {
      await expect(
        sqrtPriceMath.getPriceAfterSwap(0, expandTo18Decimals(1), expandTo18Decimals(1).div(10), false)
      ).to.be.revertedWith('SqrtPriceMath::getPriceAfterSwap: sqrtP cannot be zero')
      await expect(
        sqrtPriceMath.getPriceAfterSwap(0, expandTo18Decimals(1), expandTo18Decimals(1).div(10), true)
      ).to.be.revertedWith('SqrtPriceMath::getPriceAfterSwap: sqrtP cannot be zero')
    })

    it('fails if price is zero', async () => {
      await expect(
        sqrtPriceMath.getPriceAfterSwap(encodePriceSqrt(1, 1), 0, expandTo18Decimals(1).div(10), false)
      ).to.be.revertedWith('SqrtPriceMath::getPriceAfterSwap: liquidity cannot be zero')
      await expect(
        sqrtPriceMath.getPriceAfterSwap(encodePriceSqrt(1, 1), 0, expandTo18Decimals(1).div(10), true)
      ).to.be.revertedWith('SqrtPriceMath::getPriceAfterSwap: liquidity cannot be zero')
    })

    it('returns input price if amount in is zero', async () => {
      expect(
        await sqrtPriceMath.getPriceAfterSwap(encodePriceSqrt(1, 1), expandTo18Decimals(1).div(10), 0, false)
      ).to.eq(encodePriceSqrt(1, 1))
      expect(
        await sqrtPriceMath.getPriceAfterSwap(encodePriceSqrt(1, 1), expandTo18Decimals(1).div(10), 0, true)
      ).to.eq(encodePriceSqrt(1, 1))
    })

    it('sqrtQ is greater than sqrtP if one for zero', async () => {
      const sqrtQ = await sqrtPriceMath.getPriceAfterSwap(
        encodePriceSqrt(1, 1),
        expandTo18Decimals(1),
        expandTo18Decimals(1).div(10),
        false
      )
      expect(sqrtQ).to.be.gt(encodePriceSqrt(1, 1))
    })
    it('sqrtQ is less than sqrtP if zero for one', async () => {
      const sqrtQ = await sqrtPriceMath.getPriceAfterSwap(
        encodePriceSqrt(1, 1),
        expandTo18Decimals(1),
        expandTo18Decimals(1).div(10),
        true
      )
      expect(sqrtQ).to.be.lt(encodePriceSqrt(1, 1))
    })
    it('sqrtQ is equal to sqrtP if amount in is zero and swapping one for zero', async () => {
      const sqrtQ = await sqrtPriceMath.getPriceAfterSwap(
        encodePriceSqrt(1, 1),
        expandTo18Decimals(1),
        BigNumber.from(0),
        false
      )
      expect(sqrtQ).to.eq(encodePriceSqrt(1, 1))
    })
    it('sqrtQ is equal to sqrtP if amount in is zero and swapping zero for one', async () => {
      const sqrtQ = await sqrtPriceMath.getPriceAfterSwap(
        encodePriceSqrt(1, 1),
        expandTo18Decimals(1),
        BigNumber.from(0),
        true
      )
      expect(sqrtQ).to.eq(encodePriceSqrt(1, 1))
    })
    it('price of 1 to 1.21', async () => {
      const sqrtQ = await sqrtPriceMath.getPriceAfterSwap(
        encodePriceSqrt(1, 1),
        expandTo18Decimals(1),
        expandTo18Decimals(1).div(10),
        false
      )
      expect(sqrtQ).to.eq(encodePriceSqrt(121, 100))
    })
    it('price of 1 to 1/1.21', async () => {
      const sqrtQ = await sqrtPriceMath.getPriceAfterSwap(
        encodePriceSqrt(1, 1),
        expandTo18Decimals(1),
        expandTo18Decimals(1).div(10),
        true
      )
      expect(sqrtQ).to.eq(encodePriceSqrt(100, 121))
    })
  })
})
