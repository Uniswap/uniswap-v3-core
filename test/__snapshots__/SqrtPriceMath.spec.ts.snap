// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SqrtPriceMath #getAmountDeltas gas cost for amount0/false 1`] = `1088`;

exports[`SqrtPriceMath #getAmountDeltas gas cost for amount0/true 1`] = `1215`;

exports[`SqrtPriceMath #getAmountDeltas gas cost for amount1/false 1`] = `965`;

exports[`SqrtPriceMath #getAmountDeltas gas cost for amount1/true 1`] = `971`;

exports[`SqrtPriceMath #getNextPrice zeroForOne = false gas 1`] = `760`;

exports[`SqrtPriceMath #getNextPrice zeroForOne = true gas 1`] = `1532`;
